File: doc/code_examples/Tutorial_ProForma.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
Tutorial: ProForma v2 Parser and Writer

This tutorial demonstrates how to use the ProForma v2 parser and writer
in pyOpenMS for representing peptides with modifications.

ProForma is a standardized notation for proteoforms developed by the
HUPO Proteomics Standards Initiative (PSI).
"""

import pyopenms as oms


def basic_parsing():
    """Basic parsing and serialization examples."""
    print("=== Basic Parsing ===")

    # Parse a simple peptide
    pf = oms.parse("PEPTIDE")
    print(f"Simple: {oms.toString(pf, oms.ProFormaWriteMode.LOSSLESS)}")

    # Parse with UNIMOD modification (Oxidation on M)
    pf = oms.parse("PEM[UNIMOD:35]TIDE")
    print(f"UNIMOD: {oms.toString(pf, oms.ProFormaWriteMode.LOSSLESS)}")

    # Parse with named modification
    pf = oms.parse("PES[Phospho]TIDE")
    print(f"Named:  {oms.toString(pf, oms.ProFormaWriteMode.LOSSLESS)}")

    # Parse with mass delta
    pf = oms.parse("PEM[+15.9949]TIDE")
    print(f"Delta:  {oms.toString(pf, oms.ProFormaWriteMode.LOSSLESS)}")
    print()


def terminal_modifications():
    """Terminal modification examples."""
    print("=== Terminal Modifications ===")

    # N-terminal acetylation
    pf = oms.parse("[Acetyl]-PEPTIDE")
    print(f"N-term: {oms.toString(pf, oms.ProFormaWriteMode.LOSSLESS)}")

    # C-terminal amidation
    pf = oms.parse("PEPTIDE-[Amidated]")
    print(f"C-term: {oms.toString(pf, oms.ProFormaWriteMode.LOSSLESS)}")

    # Both terminals modified
    pf = oms.parse("[Acetyl]-PEPTIDE-[Amidated]")
    print(f"Both:   {oms.toString(pf, oms.ProFormaWriteMode.LOSSLESS)}")
    print()


def write_modes():
    """Demonstrate LOSSLESS vs CANONICAL write modes."""
    print("=== Write Modes ===")

    # Parse with limited precision mass delta
    pf = oms.parse("EM[+15.99]K")

    # Lossless preserves original formatting
    lossless = oms.toString(pf, oms.ProFormaWriteMode.LOSSLESS)
    print(f"Lossless:  {lossless}")

    # Canonical normalizes to 4 decimal places
    canonical = oms.toString(pf, oms.ProFormaWriteMode.CANONICAL)
    print(f"Canonical: {canonical}")
    print()


def advanced_features():
    """Advanced ProForma features."""
    print("=== Advanced Features ===")

    # Global isotope labeling (SILAC)
    pf = oms.parse("<13C>PEPTIDE")
    print(f"Isotope label: {oms.toString(pf, oms.ProFormaWriteMode.LOSSLESS)}")

    # Global modification on specific residues (TMT)
    pf = oms.parse("<[TMT6plex]@K,N-term>PEPTIDEK")
    print(f"TMT labeling: {oms.toString(pf, oms.ProFormaWriteMode.LOSSLESS)}")

    # Unlocalised modification (position unknown)
    pf = oms.parse("[Phospho]?STYPEPTIDE")
    print(f"Unlocalised: {oms.toString(pf, oms.ProFormaWriteMode.LOSSLESS)}")

    # Labile modification (lost during fragmentation)
    pf = oms.parse("{Glycan:Hex}PEPTIDE")
    print(f"Labile: {oms.toString(pf, oms.ProFormaWriteMode.LOSSLESS)}")
    print()


def multi_chain_and_charge():
    """Multi-chain peptides and charge states."""
    print("=== Multi-chain and Charge States ===")

    # Parse with charge state
    pfi = oms.parseIon("PEPTIDE/2")
    print(f"Charged ion - chains: {len(pfi.chains)}")

    # Multi-chain (cross-linked peptides)
    pfi = oms.parseIon("PEPTIDECK[XLMOD:02001#XL1]//AC[#XL1]DEFGH")
    print(f"Cross-linked - chains: {len(pfi.chains)}")

    # Adduct charge notation
    pfi = oms.parseIon("PEPTIDE/[Na:z+1,H:z+1]")
    print(f"Adduct notation - chains: {len(pfi.chains)}")
    print()


def aasequence_conversion():
    """Converting between ProForma and AASequence."""
    print("=== AASequence Conversion ===")

    # Parse and check if convertible
    pf = oms.parse("PEM[UNIMOD:35]TIDE")

    if oms.isRepresentableAsAASequence(pf):
        # Resolve modifications using ModificationsDB
        oms.resolveModifications(pf)

        # Convert to AASequence
        aas = oms.toAASequence(pf, oms.AASequenceConversionPolicy.STRICT)
        print(f"AASequence: {aas.toString()}")
        print(f"Mono mass:  {aas.getMonoWeight():.4f} Da")

    # Check conversion issues for complex peptidoforms
    pf = oms.parse("[Phospho]?STYPEPTIDE")
    issues = oms.getAASequenceConversionIssues(pf)
    print(f"\nConversion issues for '[Phospho]?STYPEPTIDE':")
    for issue in issues:
        print(f"  - {issue.description}")

    # Use BEST_EFFORT to convert what's possible
    aas = oms.toAASequence(pf, oms.AASequenceConversionPolicy.BEST_EFFORT)
    print(f"Best effort: {aas.toString()}")
    print()


def aasequence_to_proforma():
    """Converting AASequence to ProForma."""
    print("=== AASequence to ProForma ===")

    # Create AASequence the traditional way
    aas = oms.AASequence.fromString("PEM(Oxidation)TIDE")
    print(f"AASequence: {aas.toString()}")

    # Convert to ProForma
    pf = oms.fromAASequence(aas)
    proforma_str = oms.toString(pf, oms.ProFormaWriteMode.CANONICAL)
    print(f"ProForma:   {proforma_str}")
    print()


def json_serialization():
    """JSON serialization for storage and interchange."""
    print("=== JSON Serialization ===")
    import json

    # Serialize Peptidoform to JSON
    pf = oms.parse("[Acetyl]-PEM[UNIMOD:35]TIDE")
    json_str = oms.peptidoformToJSON(pf)
    print("Peptidoform JSON (truncated):")
    print(f"  {json_str[:80]}...")

    # Deserialize from JSON
    restored = oms.peptidoformFromJSON(json_str)
    original = oms.toString(pf, oms.ProFormaWriteMode.LOSSLESS)
    roundtrip = oms.toString(restored, oms.ProFormaWriteMode.LOSSLESS)
    print(f"Roundtrip OK: {original == roundtrip}")

    # PeptidoformIon JSON
    pfi = oms.parseIon("PEPTIDE/2")
    json_str = oms.peptidoformIonToJSON(pfi)
    restored = oms.peptidoformIonFromJSON(json_str)
    print(f"Ion roundtrip OK: {len(pfi.chains) == len(restored.chains)}")
    print()


def batch_processing():
    """Process multiple peptides and calculate masses."""
    print("=== Batch Processing ===")

    peptides = [
        "PEPTIDE",
        "PEM[UNIMOD:35]TIDE",
        "[Acetyl]-PEPTIDEK",
        "PEPS[UNIMOD:21]TIDE",
    ]

    print(f"{'ProForma':<30} {'Mass (Da)':>12}")
    print("-" * 44)

    for proforma_str in peptides:
        pf = oms.parse(proforma_str)

        if oms.isRepresentableAsAASequence(pf):
            oms.resolveModifications(pf)
            aas = oms.toAASequence(pf, oms.AASequenceConversionPolicy.STRICT)
            mass = aas.getMonoWeight()
            print(f"{proforma_str:<30} {mass:>12.4f}")
        else:
            print(f"{proforma_str:<30} {'N/A':>12}")
    print()


def error_handling():
    """Demonstrate error handling."""
    print("=== Error Handling ===")

    invalid_inputs = [
        ("[Invalid", "Unclosed bracket"),
        ("", "Empty input"),
        ("PEP[", "Incomplete modification"),
    ]

    for proforma_str, description in invalid_inputs:
        try:
            pf = oms.parse(proforma_str)
            print(f"  '{proforma_str}': Unexpectedly succeeded")
        except Exception as e:
            print(f"  {description}: Caught exception (expected)")
    print()


if __name__ == "__main__":
    print("=" * 60)
    print("ProForma v2 Tutorial for pyOpenMS")
    print("=" * 60)
    print()

    basic_parsing()
    terminal_modifications()
    write_modes()
    advanced_features()
    multi_chain_and_charge()
    aasequence_conversion()
    aasequence_to_proforma()
    json_serialization()
    batch_processing()
    error_handling()

    print("Tutorial completed successfully!")

File: doc/code_examples/Tutorial_ProForma.cpp
// Copyright (c) 2002-present, OpenMS Inc. -- EKU Tuebingen, ETH Zurich, and FU Berlin
// SPDX-License-Identifier: BSD-3-Clause
//

//! [doxygen_snippet_ProForma]

// This tutorial demonstrates how to use the ProForma v2 parser and writer
// for representing peptides with modifications.
//
// ProForma is a standardized notation for proteoforms developed by the
// HUPO Proteomics Standards Initiative (PSI).

#include <OpenMS/CHEMISTRY/ProForma.h>
#include <OpenMS/CHEMISTRY/AASequence.h>
#include <OpenMS/KERNEL/MSSpectrum.h>
#include <iostream>

using namespace OpenMS;
using namespace std;

// Type aliases for ProForma nested types
using Peptidoform = ProForma::Peptidoform;
using PeptidoformIon = ProForma::PeptidoformIon;
using WriteMode = ProForma::WriteMode;
using ConversionPolicy = ProForma::ConversionPolicy;

int main()
{
  // ===========================================
  // Basic Parsing
  // ===========================================
  cout << "=== Basic Parsing ===" << endl;

  // Parse a simple peptide
  Peptidoform pf = ProForma::parse("PEPTIDE");
  cout << "Simple: " << ProForma::toString(pf, WriteMode::LOSSLESS) << endl;

  // Parse with UNIMOD modification (Oxidation on M)
  pf = ProForma::parse("PEM[UNIMOD:35]TIDE");
  cout << "UNIMOD: " << ProForma::toString(pf, WriteMode::LOSSLESS) << endl;

  // Parse with named modification
  pf = ProForma::parse("PES[Phospho]TIDE");
  cout << "Named:  " << ProForma::toString(pf, WriteMode::LOSSLESS) << endl;

  // Parse with mass delta
  pf = ProForma::parse("PEM[+15.9949]TIDE");
  cout << "Delta:  " << ProForma::toString(pf, WriteMode::LOSSLESS) << endl;
  cout << endl;

  // ===========================================
  // Terminal Modifications
  // ===========================================
  cout << "=== Terminal Modifications ===" << endl;

  // N-terminal acetylation
  pf = ProForma::parse("[Acetyl]-PEPTIDE");
  cout << "N-term: " << ProForma::toString(pf, WriteMode::LOSSLESS) << endl;

  // C-terminal amidation
  pf = ProForma::parse("PEPTIDE-[Amidated]");
  cout << "C-term: " << ProForma::toString(pf, WriteMode::LOSSLESS) << endl;

  // Both terminals modified
  pf = ProForma::parse("[Acetyl]-PEPTIDE-[Amidated]");
  cout << "Both:   " << ProForma::toString(pf, WriteMode::LOSSLESS) << endl;
  cout << endl;

  // ===========================================
  // Write Modes: LOSSLESS vs CANONICAL
  // ===========================================
  cout << "=== Write Modes ===" << endl;

  // Parse with limited precision mass delta
  pf = ProForma::parse("EM[+15.99]K");

  // Lossless preserves original formatting
  cout << "Lossless:  " << ProForma::toString(pf, WriteMode::LOSSLESS) << endl;

  // Canonical normalizes to 4 decimal places
  cout << "Canonical: " << ProForma::toString(pf, WriteMode::CANONICAL) << endl;
  cout << endl;

  // ===========================================
  // Advanced Features
  // ===========================================
  cout << "=== Advanced Features ===" << endl;

  // Global isotope labeling (SILAC)
  pf = ProForma::parse("<13C>PEPTIDE");
  cout << "Isotope label: " << ProForma::toString(pf, WriteMode::LOSSLESS) << endl;

  // Global modification on specific residues (TMT)
  pf = ProForma::parse("<[TMT6plex]@K,N-term>PEPTIDEK");
  cout << "TMT labeling:  " << ProForma::toString(pf, WriteMode::LOSSLESS) << endl;

  // Unlocalised modification (position unknown)
  pf = ProForma::parse("[Phospho]?STYPEPTIDE");
  cout << "Unlocalised:   " << ProForma::toString(pf, WriteMode::LOSSLESS) << endl;

  // Labile modification (lost during fragmentation)
  pf = ProForma::parse("{Glycan:Hex}PEPTIDE");
  cout << "Labile:        " << ProForma::toString(pf, WriteMode::LOSSLESS) << endl;
  cout << endl;

  // ===========================================
  // Multi-chain Peptides and Charge States
  // ===========================================
  cout << "=== Multi-chain and Charge States ===" << endl;

  // Parse with charge state (returns PeptidoformIon)
  PeptidoformIon pfi = ProForma::parseIon("PEPTIDE/2");
  cout << "Charged ion - chains: " << pfi.chains.size();
  if (pfi.charge.has_value() && std::holds_alternative<int>(pfi.charge.value()))
  {
    cout << ", charge: " << std::get<int>(pfi.charge.value());
  }
  cout << endl;

  // Cross-linked peptides
  pfi = ProForma::parseIon("PEPK[+138.068#XL1]IDE//ANOK[#XL1]THER/3");
  cout << "Cross-linked - chains: " << pfi.chains.size() << endl;
  cout << endl;

  // ===========================================
  // Mass Calculation
  // ===========================================
  cout << "=== Mass Calculation ===" << endl;

  pf = ProForma::parse("PEM[UNIMOD:35]TIDE");
  ProForma::resolveModifications(pf);

  if (ProForma::canCalculateMass(pf))
  {
    double mass = ProForma::getMonoWeight(pf);
    cout << "Monoisotopic mass: " << mass << " Da" << endl;

    double mz = ProForma::getMZ(pf, 2);
    cout << "m/z at charge 2+: " << mz << endl;
  }
  cout << endl;

  // ===========================================
  // AASequence Conversion
  // ===========================================
  cout << "=== AASequence Conversion ===" << endl;

  pf = ProForma::parse("PEM[UNIMOD:35]TIDE");

  if (ProForma::isRepresentableAsAASequence(pf))
  {
    ProForma::resolveModifications(pf);
    AASequence aas = ProForma::toAASequence(pf, ConversionPolicy::FAIL_ON_LOSS);
    cout << "AASequence: " << aas.toString() << endl;
    cout << "Mono mass:  " << aas.getMonoWeight() << " Da" << endl;
  }

  // Convert AASequence back to ProForma
  AASequence aas = AASequence::fromString("PEM(Oxidation)TIDE");
  pf = ProForma::fromAASequence(aas);
  cout << "ProForma:   " << ProForma::toString(pf, WriteMode::CANONICAL) << endl;
  cout << endl;

  // ===========================================
  // Theoretical Spectrum Generation
  // ===========================================
  cout << "=== Spectrum Generation ===" << endl;

  pf = ProForma::parse("PEPTIDE");
  ProForma::resolveModifications(pf);

  if (ProForma::canGenerateSpectrum(pf))
  {
    // Generate b/y ions at charge 1-2
    MSSpectrum spec = ProForma::generateSpectrum(pf, 1, 2, "by", false, true);
    cout << "Generated spectrum with " << spec.size() << " peaks" << endl;
  }

  // Cross-linked peptide spectrum
  pfi = ProForma::parseIon("PEPK[+138.068#XL1]IDE//ANOK[#XL1]THER/3");
  ProForma::resolveModifications(pfi.chains[0]);
  ProForma::resolveModifications(pfi.chains[1]);

  if (ProForma::canGenerateSpectrum(pfi))
  {
    MSSpectrum spec = ProForma::generateSpectrum(pfi, 1, 2, "aby", false, true);
    cout << "Cross-linked spectrum with " << spec.size() << " peaks" << endl;
  }
  cout << endl;

  // ===========================================
  // JSON Serialization
  // ===========================================
  cout << "=== JSON Serialization ===" << endl;

  pf = ProForma::parse("[Acetyl]-PEM[UNIMOD:35]TIDE");
  String json_str = ProForma::toJSON(pf);
  cout << "JSON length: " << json_str.size() << " bytes" << endl;

  // Roundtrip: deserialize and compare
  Peptidoform restored = ProForma::peptidoformFromJSON(json_str);
  String original = ProForma::toString(pf, WriteMode::LOSSLESS);
  String roundtrip = ProForma::toString(restored, WriteMode::LOSSLESS);
  cout << "Roundtrip OK: " << (original == roundtrip ? "yes" : "no") << endl;
  cout << endl;

  cout << "Tutorial completed successfully!" << endl;

  return 0;
}

//! [doxygen_snippet_ProForma]
